src/react-components/auth/AuthContext.js

```diff
@@ -0,0 +1,96 @@
+import React, { createContext, useState, useEffect, useCallback } from "react";
+import PropTypes from "prop-types";
+import configs from "../../utils/configs";
+
+// TODO: We really shouldn't include these dependencies on every page. A dynamic import would work better.
+import jwtDecode from "jwt-decode";
+import AuthChannel from "../../utils/auth-channel";
+import { connectToReticulum } from "../../utils/phoenix-utils";
+
+export const AuthContext = createContext();
+
+export function AuthContextProvider({ children, store }) {
+  const signIn = useCallback(
+    async email => {
+      const authChannel = new AuthChannel(store);
+      const socket = await connectToReticulum();
+      authChannel.setSocket(socket);
+      const { authComplete } = await authChannel.startAuthentication(email);
+      await authComplete;
+
+      // TODO: Doing all of this just to determine if the user is an admin seems unnecessary. The auth callback should have the isAdmin flag.
+      const retPhxChannel = socket.channel("ret", { hub_id: "index", token: store.state.credentials.token });
+
+      const perms = await new Promise(resolve => {
+        retPhxChannel.join().receive("ok", () => {
+          retPhxChannel.push("refresh_perms_token").receive("ok", ({ perms_token }) => {
+            const perms = jwtDecode(perms_token);
+            retPhxChannel.leave();
+            resolve(perms);
+          });
+        });
+      });
+
+      configs.setIsAdmin(perms.postgrest_role === "ret_admin");
+      store.update({ credentials: { isAdmin: perms.postgrest_role === "ret_admin" } });
```

Comment:
I don't think we should block the page load to check if the user is an admin especially since a majority of users of Hubs are not admins. Instead we should get this isAdmin flag into the JWT session. We should make JWT sessions expire relatively quickly and use [refresh tokens](https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/) in order to ensure that changes to profiles take effect in a reasonable period of time. Then, since we will be storing this JWT access token in localStorage, you will be able to check whether the user is an admin without blocking the page load.

That's out of scope for this PR though.

I will revert to blocking the page for now. However, the `Hubs.store.isAdmin()` method that I use in the Hubs API PR makes use of this and will need to be changed. I would like to resolve this before merging the API PR.

Reply:
I agree we should not block page loading, you just don't show the admin button right away.